File: _includes/sort-benchmark/algorithms/radix.rs
fn counting_sort_by_digit(a: &mut [usize], exp: usize) {
    let n = a.len();
    let mut output = vec![0usize; n];
    let mut count = [0usize; 10];

    for &x in a.iter() {
        count[(x / exp) % 10] += 1;
    }

    for i in 1..10 {
        count[i] += count[i - 1];
    }

    for i in (0..n).rev() {
        let digit = (a[i] / exp) % 10;
        count[digit] -= 1;
        output[count[digit]] = a[i];
    }

    a.copy_from_slice(&output);
}

fn radix_sort(a: &mut [usize]) {
    if a.is_empty() {
        return;
    }

    let max = *a.iter().max().unwrap();
    let mut exp = 1usize;

    while max / exp > 0 {
        counting_sort_by_digit(a, exp);
        exp *= 10;
    }
}

File: _includes/sort-benchmark/algorithms/radix.swift
func counting_sort_by_digit(_ a: UnsafeMutableBufferPointer<Int>, _ exp: Int) {
    let n = a.count
    var output = [Int](repeating: 0, count: n)
    var count = [Int](repeating: 0, count: 10)

    for i in 0..<n {
        count[(a[i] / exp) % 10] += 1
    }

    for i in 1..<10 {
        count[i] += count[i - 1]
    }

    for i in (0..<n).reversed() {
        let digit = (a[i] / exp) % 10
        count[digit] -= 1
        output[count[digit]] = a[i]
    }

    for i in 0..<n {
        a[i] = output[i]
    }
}

func radix_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { radix_sort($0) }
}

func radix_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.isEmpty {
        return
    }

    var max = a[0]
    for i in 1..<a.count {
        if a[i] > max {
            max = a[i]
        }
    }
    var exp = 1

    while max / exp > 0 {
        counting_sort_by_digit(a, exp)
        exp *= 10
    }
}
